File: backend/api/notifications.py
from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from typing import List, Optional
from utils.supabase_client import supabase
from services.notification_service import NotificationService

router = APIRouter()

class PreferenceUpdate(BaseModel):
    email_enabled: Optional[bool] = None
    push_enabled: Optional[bool] = None
    whatsapp_enabled: Optional[bool] = None
    sms_enabled: Optional[bool] = None
    quiet_start: Optional[str] = None  # HH:MM format
    quiet_end: Optional[str] = None
    preferred_days: Optional[List[int]] = None
    preferred_time: Optional[str] = None
    max_items_per_notification: Optional[int] = None
    snoozed_until: Optional[str] = None

@router.get("/notifications/preferences", response_model=dict)
async def get_preferences(user_id: str):
    resp = supabase.table("notification_preferences").select("*").eq("user_id", user_id).single().execute()
    if resp.error:
        raise HTTPException(status_code=404, detail="Preferences not found")
    return resp.data

@router.post("/notifications/preferences", response_model=dict)
async def set_preferences(user_id: str, prefs: PreferenceUpdate):
    # Upsert preference record
    data = {k: v for k, v in prefs.dict().items() if v is not None}
    data["user_id"] = user_id
    resp = supabase.table("notification_preferences").upsert(data, on_conflict=["user_id"]).execute()
    if resp.error:
        raise HTTPException(status_code=400, detail=resp.error.message)
    return resp.data[0] if isinstance(resp.data, list) else resp.data

@router.post("/notifications/subscribe", status_code=status.HTTP_201_CREATED)
async def subscribe_push(user_id: str, subscription: dict):
    # Store VAPID subscription JSON
    resp = supabase.table("push_subscriptions").upsert({"user_id": user_id, "subscription": subscription}, on_conflict=["user_id"]).execute()
    if resp.error:
        raise HTTPException(status_code=400, detail=resp.error.message)
    return {"result": "subscribed"}

@router.post("/notifications/unsubscribe", status_code=status.HTTP_200_OK)
async def unsubscribe_push(user_id: str):
    resp = supabase.table("push_subscriptions").delete().eq("user_id", user_id).execute()
    if resp.error:
        raise HTTPException(status_code=400, detail=resp.error.message)
    return {"result": "unsubscribed"}

@router.post("/notifications/send-test", status_code=status.HTTP_200_OK)
async def send_test(user_id: str):
    service = NotificationService()
    pref = service.get_user_preferences(user_id)
    if not pref.email_enabled:
        raise HTTPException(status_code=400, detail="Email not enabled for user")
    service.send_email(to_email=user_id, subject="QueueIt Test Notification", html_content="<p>This is a test notification from QueueIt.</p>")
    return {"result": "test email sent"}

File: backend_node/src/routes/notifications.ts
import { Router, Response } from 'express';
import { requireAuth, AuthenticatedRequest } from '../middleware/auth';
import { supabase } from '../config/supabase';
import { NotificationService } from '../services/notificationService';

const router = Router();

function getUserId(req: AuthenticatedRequest): string {
  return (req.user?.id || req.user?.sub || '') as string;
}

// ---------------------------------------------------------------------------
// GET /api/notifications/preferences
// ---------------------------------------------------------------------------
router.get('/preferences', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  const userId = getUserId(req);
  try {
    const { data, error } = await supabase.from('notification_preferences').select('*').eq('user_id', userId).maybeSingle();
    if (error || !data) {
      return res.status(404).json({ detail: 'Preferences not found' });
    }
    return res.json(data);
  } catch (err: any) {
    return res.status(400).json({ detail: err.message || String(err) });
  }
});

// ---------------------------------------------------------------------------
// POST /api/notifications/preferences
// ---------------------------------------------------------------------------
router.post('/preferences', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  const userId = getUserId(req);
  const prefs = req.body || {};
  try {
    const updateData: Record<string, any> = { user_id: userId };
    Object.keys(prefs).forEach((k) => {
      if (prefs[k] !== undefined && k !== 'user_id') {
        updateData[k] = prefs[k];
      }
    });

    const { data, error } = await supabase.from('notification_preferences').upsert(updateData, { onConflict: 'user_id' }).select();
    if (error) {
      return res.status(400).json({ detail: error.message });
    }
    return res.json(data?.[0] || updateData);
  } catch (err: any) {
    return res.status(400).json({ detail: err.message || String(err) });
  }
});

// ---------------------------------------------------------------------------
// POST /api/notifications/subscribe
// ---------------------------------------------------------------------------
router.post('/subscribe', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  const userId = getUserId(req);
  const { subscription } = req.body || {};
  try {
    const { error } = await supabase.from('push_subscriptions').upsert({ user_id: userId, subscription }, { onConflict: 'user_id' });
    if (error) {
      return res.status(400).json({ detail: error.message });
    }
    return res.status(201).json({ result: 'subscribed' });
  } catch (err: any) {
    return res.status(400).json({ detail: err.message || String(err) });
  }
});

// ---------------------------------------------------------------------------
// POST /api/notifications/unsubscribe
// ---------------------------------------------------------------------------
router.post('/unsubscribe', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  const userId = getUserId(req);
  try {
    const { error } = await supabase.from('push_subscriptions').delete().eq('user_id', userId);
    if (error) {
      return res.status(400).json({ detail: error.message });
    }
    return res.json({ result: 'unsubscribed' });
  } catch (err: any) {
    return res.status(400).json({ detail: err.message || String(err) });
  }
});

// ---------------------------------------------------------------------------
// POST /api/notifications/send-test
// ---------------------------------------------------------------------------
router.post('/send-test', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  const userId = getUserId(req);
  try {
    const { data: pref } = await supabase.from('notification_preferences').select('*').eq('user_id', userId).maybeSingle();
    if (pref && pref.email_enabled === false) {
      return res.status(400).json({ detail: 'Email not enabled for user' });
    }
    const service = new NotificationService();
    const userEmail = req.user?.email || userId;
    await service.sendEmailAsync(userEmail, 'QueueIt Test Notification', '<p>This is a test notification from QueueIt.</p>');
    return res.json({ result: 'test email sent' });
  } catch (err: any) {
    return res.status(400).json({ detail: err.message || String(err) });
  }
});

export default router;
